java/arcs/core/allocator/Arc.kt

```diff
@@ -91,56 +90,62 @@ class Arc internal constructor(
     /** Called whenever the [ArcState] changes to [Error]. */
     fun onError(handler: () -> Unit) = onArcStateChangeFiltered(Error, handler)
 
-    private fun fireArcStateChange() {
-        sync(this) {
-            arcStateChangeHandlers.forEach {
-                it(arcState)
-            }
-        }
+    private fun fireArcStateChange(handlers: List<(ArcState) -> Unit>) = handlers.forEach {
+        it(arcState)
     }
 
-    private fun recomputeArcState() {
-        sync(this) {
-            val states = arcStatesByHostId.values
-            arcState = when {
+    private fun recomputeArcState(): List<(ArcState) -> Unit> {
+        val states = arcStatesByHostId.value.values.toList()
+        val oldState = arcStateInternal.value
+        val newState = when {
                 states.any { it == Deleted } -> Deleted
                 states.any { it == Error } -> Error
                 states.all { it == Running } -> Running
                 states.all { it == Stopped } -> Stopped
                 states.all { it == NeverStarted } -> NeverStarted
                 else -> Indeterminate
             }
+        arcStateInternal.update { newState }
```

Comment:
Given the `newState` is inferred by the current `states` volatile value.
Is there a case where there are two racing `states`: one is resolved as Running and the other as non-Running, the non-Running overwrites the Running or vice versa turns out hanging at an unexpected state forever as one of them might be never observed?

Reply:
I think the fire() will still happen on each incoming state change, but the fire() method reads the arcStateInternal.value, which could have been updated between the time recomputeArcState() is called and the fire() method reads it, so I either need to return it, or invoke fire() from here. 

I think as long as we ensure that if recomputeArcState() computes a changed state, it fires that change, every change will be observed? I'll push a fix for this. I'll try to think of a test that exercise racing behavior and run it x10000.

Reply:
What if two recomputeArcState() happens at the same time (one from waitFor() and the other from state change callback at archost)?
--timeline--
#T1: newState == Non-Running                      arcStateInternal.update { Non-Running }
#T2:       newState == Running        arcStateInternal.update { Running }  

If using the arcState directly, the Running state will never be observed.
And yes, we should fire the newState.

Reply:
Ah, btw, not only fire() the newState but also notice the current arcState will be kept at Non-Running state until next fetchCurrentStates-and-recomputeArcState kicks in.

Reply:
Given all the back-and-forth discussion.. I'll reiterate that it'd be hugely valuable to have a test case which forces the race to occur - to guarantee that the solution in this PR fixes it (and that we don't regress).

Reply:
> Given all the back-and-forth discussion.. I'll reiterate that it'd be hugely valuable to have a test case which forces the race to occur - to guarantee that the solution in this PR fixes it (and that we don't regress).

@cromwellian and I did talk about testing, he was planning to look at adding a test. The unfortunate reality of the nature of these things is the test has to be written so carefully instrumented to allow blocking in the right places to repro the theoretical race conditions, via latches and so forth, that it can make the code itself fragile or introduce issues accidentally. But it seemed worthwhile as you note, at least a first pass for review/discussion.

A higher level fuzzing-style way to pound on things Arc-lifecycle-wise, like an evolution of @ianull STS stability to operate at the next level up, could turn out to be more valuable and broadly applicable.

Reply:
I think it might be better if I go back to my original plan of using Flows, at the cost of waitFor() launching a co-routine to collect. This PR is taking longer than I thought, and the amount of code to force a race is larger than the fix. :)